### ISSUE TITLE
[typescript] Use discriminated unions for `MessageRequestResponse`

### ISSUE THREAD

**Issue by baptistemarchand** [Vonage/vonage-node-sdk#353]
The TypeScript typing for `MessageRequestResponse` could be improved.

Here is the current typing:
```typescript
    export interface MessageRequestResponseSuccess {
        status: MessageRequestResponseStatusCode,
        // some other fields...
    }
    export interface MessageError {
        status: MessageRequestResponseStatusCode;
        // some other fields
    }
    export interface MessageRequestResponse {
      'message-count': number;
      messages: (MessageRequestResponseSuccess | MessageError)[];
    }
```

So when I get a message the type is `MessageRequestResponseSuccess | MessageError` and TypeScript doesn't let me access the fields in either types, only the `status` field because it is present in both types.
For example:
```typescript
const response = await nexmo.message.sendSms(...)
for (const message of response.messages) {
      if (message.status === '0') {
        const messageId = message['message-id'] // ERROR HERE: Property 'message-id' does not exist on type 'MessageRequestResponseSuccess | MessageError'
      }
}
```
To fix this you just have to have a field that is present in both `MessageRequestResponseSuccess` and `MessageError` (like `status`) but which has a different type. For example:
```typescript
    export interface MessageRequestResponseSuccess {
        success: true
        // some other fields...
    }

    export interface MessageError {
        success: false
        // some other fields
    }

// now we can do
for (const message of response.messages) {
      if (message.success === true) {
        const messageId = message['message-id'] // no error
      }
}
```

This is just a simple example of how you can solve this problem but there are many other options. The best would probably be to use the `status` field for this. Something like that:
```typescript
    export interface MessageRequestResponseSuccess {
        status: '0'
        // some other fields...
    }

    export interface MessageError {
        status: '1' | '2' | ...
        // some other fields
    }
``` 

Thanks!

**Comment by kellyjandrews** [comment]
Good info - much appreciated.  We currently offer very minimal support for Typescript, but I'm working on a new major version which will address more of these issues.

**Comment by danielglazer** [comment]
@kellyjandrews I understand that currently your typescript support is minimal and I would like to help fix that starting with this issue if possible :) what @baptistemarchand mentioned as a possible solution is 'A common technique' according to the [ts-handbook](https://www.typescriptlang.org/docs/handbook/unions-and-intersections.html#discriminating-unions). Please update me if I can make PR for this issue

**Comment by kellyjandrews** [comment]
@danielglazer If you would like to submit something, that would be great. My main priority is writing a new version that moves types directly into the code base and sorts out most of these problems - but it would help v2. Appreciate any help you can offer.

**Comment by danielglazer** [comment]
I think that this issue can be closed now since the #398 PR was merged. Thanks, @kellyjandrews
